core/src/test/scala/org/apache/spark/sql/delta/DeltaSuite.scala

```diff
@@ -633,7 +636,53 @@ class DeltaSuite extends QueryTest
         .mode("overwrite")
         .option("partitionOverwriteMode", "dynamic")
         .save(tempDir.getCanonicalPath)
-      checkDatasetUnorderly(data.toDF.select($"value".as[String]), "a", "b", "d")
+      checkDatasetUnorderly(data.toDF.select("value").as[String], "a", "b", "d")
+    }
+
+    withTempDir { tempDir =>
+      // overwrites nothing
+
+      def data: DataFrame = spark.read.format("delta").load(tempDir.toString)
+
+      Seq(("a", "x"), ("b", "y"), ("c", "x")).toDF("value", "part")
+        .write
+        .format("delta")
+        .partitionBy("part")
+        .mode("append")
+        .save(tempDir.getCanonicalPath)
+
+      Seq(("d", "z")).toDF("value", "part")
+        .write
+        .format("delta")
+        .partitionBy("part")
+        .mode("overwrite")
+        .option("partitionOverwriteMode", "dynamic")
+        .save(tempDir.getCanonicalPath)
+      checkDatasetUnorderly(data.select("value", "part").as[(String, String)],
+        ("a", "x"), ("b", "y"), ("c", "x"), ("d", "z"))
+    }
+
+    withTempDir { tempDir =>
+      // multiple partition columns
+
+      def data: DataFrame = spark.read.format("delta").load(tempDir.toString)
+
+      Seq(("a", "x", 1), ("b", "y", 2), ("c", "x", 3)).toDF("part1", "part2", "value")
+        .write
+        .format("delta")
+        .partitionBy("part1", "part2")
+        .mode("append")
+        .save(tempDir.getCanonicalPath)
+
+      Seq(("x", "a", 4), ("x", "d", 5)).toDF("part2", "part1", "value")
```

Comment:
nit: why did you switch the column order here? needlessly confusing

Reply:
i was overly ambitious in testing things. reverted.